pkg/server/local_observer_test.go

```diff
@@ -114,3 +109,71 @@ func TestLocalObserverServer_GetFlows(t *testing.T) {
 	assert.NoError(t, err)
 	assert.Equal(t, req.Number, uint64(i))
 }
+
+type testHook struct {
+	t            *testing.T
+	ciliumDaemon interface{}
+	stopEvery    int64
+
+	seenFlows int64
+}
+
+func (h *testHook) AfterServerInit(srv serveroption.Server) error {
+	assert.Equal(h.t, srv.GetOptions().CiliumDaemon, h.ciliumDaemon)
+	return nil
+}
+
+func (h *testHook) BeforeDecoding(ctx context.Context, payload *pb.Payload) (stop bool, err error) {
+	h.seenFlows++
+	assert.Equal(h.t, payload.Time.Seconds, h.seenFlows-1)
+	if h.seenFlows%h.stopEvery == 0 {
+		stop = true
+	}
+	return stop, nil
+}
+
+func (h *testHook) AfterDecoding(ctx context.Context, event *v1.Event) (stop bool, err error) {
+	assert.Equal(h.t, event.Timestamp.Seconds, h.seenFlows-1)
+	if h.seenFlows%h.stopEvery == 0 {
+		assert.Fail(h.t, "server did not stop decoding")
+	}
+	return stop, nil
+}
+
+func TestHooks(t *testing.T) {
+	numFlows := 10
+	queueSize := 0
+
+	type fakeCiliumDaemon struct{}
+	h := &testHook{
+		t:            t,
+		stopEvery:    2,
+		ciliumDaemon: &fakeCiliumDaemon{},
+	}
+
+	pp := noopParser(t)
+	s, err := NewLocalServer(pp, logger.GetLogger(),
+		serveroption.WithMaxFlows(numFlows),
+		serveroption.WithMonitorBuffer(queueSize),
+		serveroption.WithCiliumDaemon(h.ciliumDaemon),
+		serveroption.WithAfterServerInit(h),
+		serveroption.WithBeforeDecoding(h),
+		serveroption.WithAfterDecoding(h))
+	require.NoError(t, err)
+	go s.Start()
+
+	m := s.GetEventsChannel()
+	for i := 0; i < numFlows; i++ {
+		tn := monitor.TraceNotifyV0{Type: byte(monitorAPI.MessageTypeTrace)}
+		data := testutils.MustCreateL3L4Payload(tn)
+		pl := &pb.Payload{
+			Time: &types.Timestamp{Seconds: int64(i)},
+			Type: pb.EventType_EventSample,
+			Data: data,
+		}
+		m <- pl
+	}
+	close(s.GetEventsChannel())
+	<-s.GetStopped()
+	assert.Equal(t, h.seenFlows, int64(numFlows))
```

Comment:
assertion is better done in the other direction -- first is the value you expect, the other is the result of running the test.

This will generate a much better error message

Reply:
Good catch!